sbncode/Calibration/TrackCaloSkimmer_module.cc

```diff
@@ -0,0 +1,460 @@
+
+static int N_FIT_DATA = -1;
+static double FIT_RR[MAX_N_FIT_DATA];
+static double FIT_DQDX[MAX_N_FIT_DATA];
+
+void ConstResiduals(int &npar, double *g, double &result, double *par, int flag) {
+  double ret = 0;
+
+  double C = *par;
+
+  for (int i = 0; i < N_FIT_DATA; i++) {
+    double diff = FIT_DQDX[i] - C;
+    ret += diff*diff;
+  }
+
+  result = sqrt(ret);
+}
+
+void ExpResiduals(int &npar, double *g, double &result, double *par, int flag) {
+  double ret = 0;
+
+  double A = par[0];
+  double R = par[1];
+
+  for (int i = 0; i < N_FIT_DATA; i++) {
+    double diff = FIT_DQDX[i] - A*exp(-FIT_RR[i]/R);
+    ret += diff*diff;
+  }
+
+  result = sqrt(ret);
+}
+
+sbn::TrackCaloSkimmer::~TrackCaloSkimmer() {
+  delete fTrack;
+}
+
+sbn::TrackCaloSkimmer::TrackCaloSkimmer(fhicl::ParameterSet const& p)
+  : EDAnalyzer{p},
+    fFitExp(2),
+    fFitConst(1)
+{
+  // Grab config
+  fPFPproducer  = p.get< art::InputTag > ("PFPproducer","pandoraGausCryo0");
+  fT0Producer   = p.get< art::InputTag > ("T0producer", "pandoraGausCryo0");
+  fCALOproducer = p.get< art::InputTag > ("CALOproducer");
+  fTRKproducer  = p.get< art::InputTag > ("TRKproducer" );
+  fRequireT0 = p.get<bool>("RequireT0", false);
+  fDoTailFit = p.get<bool>("DoTailFit", true);
+  fHitRawDigitsTickCollectWidth = p.get<double>("HitRawDigitsTickCollectWidth", 50.);
+
+  fRawDigitproducers = p.get<std::vector<art::InputTag>>("RawDigitproducers", {});
+
+  std::vector<fhicl::ParameterSet> selection_tool_configs(p.get<std::vector<fhicl::ParameterSet>>("SelectionTools"), {});
+  for (const fhicl::ParameterSet &p: selection_tool_configs) {
+    fSelectionTools.push_back(art::make_tool<sbn::ITCSSelectionTool>(p));
+  }
+
+  // Setup meta info
+  fMeta.iproc = -1;
+  fMeta.ifile = -1;
+  const char *process_str = std::getenv("PROCESS");
+  if (process_str) {
+    try {
+      fMeta.iproc = std::stoi(process_str);
+    }
+    catch (...) {}
+  }
+
+  // Output stuff
+  fTrack = new sbn::TrackInfo();
+
+  art::ServiceHandle<art::TFileService> tfs;
+  fTree = tfs->make<TTree>("TrackCaloSkim", "Calo Tree");
+  fTree->Branch("trk", &fTrack);
+}
+
+void sbn::TrackCaloSkimmer::analyze(art::Event const& e)
+{
+  unsigned evt = e.event();
+  unsigned sub = e.subRun();
+  unsigned run = e.run();
+  std::cout << "[TrackCaloSkimmer::analyzeEvent] Run: " << run << ", SubRun: " << sub << ", Event: "<< evt << ", Is Data: " << e.isRealData() << std::endl;
+
+  fMeta.evt = evt;
+  fMeta.subrun = sub;
+  fMeta.run = run;
+
+  std::vector<art::Ptr<recob::PFParticle>> PFParticleList;
+  try {
+    art::ValidHandle<std::vector<recob::PFParticle>> pfparticles = e.getValidHandle<std::vector<recob::PFParticle>>(fPFPproducer);
+    art::fill_ptr_vector(PFParticleList, pfparticles);
+  }
+  catch(...) {
+      std::cout << "PFP's with tag: " << fPFPproducer << " not present.\n";
+      // Real data may have missing products -- just ignore the event
+      if (e.isRealData()) return;
```

Comment:
Do we really want this to fail silently if there is no data? I think we probably do want it to crash in that case...right?

Reply:
For the way ICARUS data has been processed, there are art::Events that have missing data products (if the reconstruction doesn't get run). So this is required functionality for processing such files.

Reply:
OK, I think this needs some followup with @SFBayLaser  and ICARUS team on cases where reco doesn't get run/missing data products. I've made little conditions/things like this in the past and it's always come back to bite me, as it inevitably hides problems.

If there's a valid reason that we won't have these objects on some events, ok. But reco failing probably means the whole job should fail and we should never get to this point.

Reply:
I don't believe it means the reco is failing. From what I can tell, there is an optical filter that runs and the reconstruction only gets run if the optical filter passes. If the filter fails, then the raw data is still there but the reconstruction does not get run.

Reply:
art filters are a total unintuitive pain. Way out of scope here, but I kind of think the reco should be putting an explicit "failed optical filter" flag in the event, and the downstream recos should be seeing that and, say, putting an empty list of tracks. Then downstream can distinguish between mysteriously missing products and the actual expected program flow.

Reply:
Not sure I agree @cjbackhouse , but that's a good and different conversation.

@gputnam if that's the case, then the FCL should be configured such that the analysis module only runs on events that pass that filter, using the `SelectEvents: ["name_of_trigger_path"]` in the analysis module configuration (like we do for output streams). I would still remove this more silent failure here.

Reply:
Isn't that only possible to do if the analysis job is in the same larsoft job as the filtering? What if I run on the output of stage0/stage1 processing with the analysis module?